File: kokobanana.go
package main

import (
	"fmt"
	"math"
)

func minEatingSpeed(piles []int, h int) int {
	maxNum := piles[0]
	pilesFloat := make([]float64, len(piles))
	for i := 0; i < len(piles); i++ {
		maxNum = max(maxNum, piles[i])
		pilesFloat[i] = float64(piles[i])
	}

	start := 0
	end := maxNum

	for start <= end {
		mid := start + (end-start)/2

		if isEatable(pilesFloat, h, float64(mid)) {
			end = mid - 1
			maxNum = mid
		} else {
			start = mid + 1
		}
	}
	return maxNum
}

func isEatable(piles []float64, h int, speed float64) bool {

	i := 0

	for i < len(piles) && h > 0 {
		hoursConsumed := int(math.Ceil(piles[i] / speed))
		h -= hoursConsumed
		i++
	}

	if i == len(piles) && h >= 0 {
		return true
	}

	return false

}

func main() {
	piles := []int{3, 6, 7, 11}
	h := 8

	fmt.Println(minEatingSpeed(piles, h))
}

File: Kokobanana.java
public class Kokobanana {
    public int minEatingSpeed(int[] piles, int h) {
        int max = piles[0];
        double[] pilesDouble = new double[piles.length];
        for(int i=0; i< piles.length; i++){
            pilesDouble[i] = piles[i];
            max = Math.max(piles[i], max);
        }

        int start = 0;
        int end = max;
        while(start <= end){
            int mid = start + (end - start )/2;

            if (isSpeedEnough(pilesDouble, h, mid)){
                end = mid-1;
                max = mid;
            } else{
                start = mid + 1;
            }
        }
        return max;
    }

    private boolean isSpeedEnough(double[] piles, int h, double speed){
        int i =0;
        while(i<piles.length && h > 0){
            int hoursConsumed = (int)Math.ceil(piles[i]/speed);
            h -= hoursConsumed;
            i++;
        }

        if (i == piles.length && h>=0){
            return true;
        }
        return false;
    }

    public static void main(String[] args) {
        int[] piles = {3,6,7,11};
        int h = 8;
        System.out.println(new Kokobanana().minEatingSpeed(piles, h));
    }
}
